### ISSUE TITLE
Refactoring Proposal

### ISSUE THREAD

**Issue by faceyspacey** [timbrandin/blaze-react#32]
Hey Tim, here's a proposal I put together of what we can do to organize this better to make it easier for new developers to contribute:

https://gist.github.com/faceyspacey/e7498f808876548f394d

Here's a public to do list comprising it:
https://docs.google.com/spreadsheets/d/1q6d4VEv6hnZlwH3Fi_jbTYAp1c50A9srMguijvCbedg/edit#gid=0

It's all based on the less formal and disorganized exploration here:
https://github.com/timbrandin/react-templates/pull/1/files#r44759084

Basically you can skip that if you want and move directly to the gist and spreadsheet.

**Comment by timbrandin** [comment]
I've started working on your list! :+1:

**Comment by timbrandin** [comment]
Hey, I've started testing a new idea of moving events parsing onto the client, here's what I have after a weekend of finding the best way to remove the need for removing the `ecmascript` module:
https://github.com/timbrandin/blaze-react/tree/no-eventmap-parsing

I don't really like this approach but there's no way of hooking into the ecmascript-compiler and reading the values of the eventsmap before our html compiling starts, I just wished that @sashko could enable such things, or allow for multiple .js compilers (I just need read-only).

**Comment by faceyspacey** [comment]
Combine your events compilation code with what the ecmascript plugin does itself, i.e. That way developers only include Blaze React, and are instructed not to include Ecmascript as well. no big deal. ES5 continues to compile just fine even if they aren't using ES6. I do exactly this in the Ecmascript package of my Meteor framework, Ultimate MVC:

https://github.com/ultimate-ide/ultimate-mvc/tree/master/lib/ultimate_ecmascript

No need to imply your compiler either:
https://github.com/ultimate-ide/ultimate-mvc/blob/master/package.js#L397

it just works if your package is added. 

...it's gonna be important the transpiled code has the events. The reason being that the natural evolution of this product--provided its successful--is that people will wanna keep the generated code instead of transpiling any time. People will end up throwing out their Blaze code, and using what we generate for them. We should be happy to help them. 

So the event handlers should be tacked on to the React component class:

```
class MyComponents extends React.Component {
     ['click .bla']() {
         //some code
    ]
}
```

People will want to inspect the generated component and see a clean comparison to the original code. That component should have the helpers, events handlers and lifecycle handlers attached to it. Tack it on the component my friend. That is in fact exactly what me and @mitar's package has. It's how React Components are structured. Proxying through `createElement` and other such hacks will obscure what's really going on from the developers. That said, I noticed you still require the developer to create the render method. You did A LOT of solid work over the weekend--so maybe that's a progression--but the idea being, besides `ReactTemplate` being unnecessary when `Template` will have the render function, is if you just tack the render method directly on to the class instead of this:

https://github.com/timbrandin/blaze-react/blob/devel/lib/exports.js#L13

we can generate a file that has the entire component looking very clean for the developer. ..Actually, I'm wrong, under the current plan with `createFromBlaze` you cant see everything in one file--since the Render method is passed from the template regex replace. ...hmm, basically we are gonna wanna move that direction. Imagine running your transpiler like via the shell, and now all your code is the React version for you to manipulate. Bye Bye your Blaze code. I can imagine that being very popular, even if most developers keep transpiling as a build step. Why? They will want to see what code we generate from them and learn how their Blaze code looks like in React. At the very least they will wanna inspect files in Chrome. ..So what I say we do is continue down the current refactoring path, i.e. moving as much into 

 and we pluck that file out, then you can look at a raw React component. ..Actually, 
Combine your events compilation code with what the ecmascript plugin does itself, i.e. That way developers only include Blaze React, and are instructed not to include Ecmascript as well. no big deal. ES5 continues to compile just fine even if they aren't using ES6. I do exactly this in the Ecmascript package of my Meteor framework, Ultimate MVC:

https://github.com/ultimate-ide/ultimate-mvc/tree/master/lib/ultimate_ecmascript

No need to imply your compiler either:
https://github.com/ultimate-ide/ultimate-mvc/blob/master/package.js#L397

it just works if your package is added. 

...it's gonna be important the transpiled code has the events. The reason being that the natural evolution of this product--provided its successful--is that people will wanna keep the generated code instead of transpiling any time. People will end up throwing out their Blaze code, and using what we generate for them. We should be happy to help them. 

So the event handlers should be tacked on to the React component class:

```
class MyComponents extends React.Component {
     ['click .bla']() {
         //some code
    ]
}
```

People will want to inspect the generated component and see a clean comparison to the original code. That component should have the helpers, events handlers and lifecycle handlers attached to it. Tack it on the component my friend. That is in fact exactly what me and @mitar's package has. It's how React Components are structured. Proxying through `createElement` and other such hacks will obscure what's really going on from the developers. That said, I noticed you still require the developer to create the render method. You did A LOT of solid work over the weekend--so maybe that's a progression--but the idea being, besides `ReactTemplate` being unnecessary when `Template` will have the render function, is if you just tack the render method directly on to the class instead of this:

https://github.com/timbrandin/blaze-react/blob/devel/lib/exports.js#L13

we can generate a file that has the entire component looking very clean for the developer. ..Actually, I'm wrong, under the current plan with `createFromBlaze` you cant see everything in one file--since the Render method is passed from the template regex replace. ...hmm, basically we are gonna wanna move that direction. Imagine running your transpiler like via the shell, and now all your code is the React version for you to manipulate. Bye Bye your Blaze code. I can imagine that being very popular, even if most developers keep transpiling as a build step. Why? They will want to see what code we generate from them and learn how their Blaze code looks like in React. At the very least they will wanna inspect files in Chrome. ..So what I say we do is continue down the current refactoring path, i.e. moving as much into `createFromBlaze` and the upcoming React mixin, but plan to down the line rewrite to generate pretty little class files. The idea being it will complicate things now, and it's important we have it as clean as possible while we think about how to accomplish things, but by the time its done we do what needs to be done in terms of "marketability." I say "marketability" because it will do wonders to be able to write readmes like your current readmes that compares your code to what's generated, and have the generated code be as close to actual React code as possible. It will really help developers migrate to React and see that it's not all that different from Blaze, and as a result quell a lot of the concerns in the Meteor forum from developers who simply are scared of non-blaze code. 

...that all said, ur idea did make me think how if we were to use this approach or a similar client runtime approach to get event selectors that can hook into child templates, it might be worth it. Basically, if we can hook into write before React renders the component, and see all its html, and simultaneously have access to the events from parent components (via React *context*), then we could attach the event to the appropriate dynamic element at runtime. I.e. cuz their classes could be changing (e.g. `active` could be added, etc). We can also keep track of parent selectors. Now that said, your current way of using cheerio to statically compile selectors won't even work, regardless if we dont support parent component selectors in v1--cuz of the example i just said regarding the `active` class being dynamically added. And that's a super common pattern for buttons. Sometimes you can't even click on a button until it has a certain class, and the selector for that class is what has the main event handler attached to it. In short, we may have no choice but to do the super dynamic runtime inherited selectors approach. 

Nevertheless, lets not worry about it. Events should be one of the last things. What you already have works just fine for the base case, which is still like 60% of all usage probably. It's enough to validate the toolset. After the rest of spacebars works, we should revisit events, and likely do the super robust approach, and do so by passing the event handlers down via the simple and straightforward `context` feature which matches up perfectly with what we're trying to do here: https://facebook.github.io/react/docs/context.html ...Then all we gotta do is find a way in `componentWillMount` to attach it, OR we do that other approach using standard `window.addEventListener`. Either way, it just seems like something that shouldn't get in the way of getting the foundation on point, especially since it's likely an area that requires serious dedication to pick the correct solution, and do it once. Look at the Velocity testing team (xolvio) they tried like 3-4 approaches to build a meteor-specific testing environment, and still couldn't get it to behave--so they put it back in Meteor's hands. Let's investigate events super deeply together at the end. Ur current event solution--transpiled on the server--is the perfect card in this house of cards to stand up for a good while while we get the other cards standing.

**Comment by faceyspacey** [comment]
one other thing: since ES5 code will compile fine even with the babel ES6 compiler, we dont need to write a separate package if we combine your js plugin code with the Ecmascript package's plugin code. However, we likely want to support @eXon's webpack meteor-webpack package, as that's likely to be a very popular thing for react/meteor developers. So we'll have to create a second version of Blaze React that combines our compiler code with the webpack compiler code. I don't see Meteor building support for multiple compilers for the same extension any time soon, given they just made their new Plugin api 2 months ago. Even if this feature is "coming soon", it's trivial to do this sorta thing, and we shouldn't wait any amount of time for MDG. 

So at the end of the project, we make 2 packages of this code. The base one with all our code, *Blaze React Base*, and then the parent ones that includes it, *Blaze React* and *Blaze React Webpack*. The parent ones have the the babel + webpack compilation related code. 


ps. @eXon's webpack compilation code might be interesting to look at and learn a thing or 2:

https://github.com/thereactivestack/meteor-webpack/blob/master/plugin/WebpackCompiler.js

**Comment by eXon** [comment]
@faceyspacey I'm not sure how this can be plugged with webpack. It makes more sense to me to use the `BlazeToReact` package (transform Blaze template into React component) than the opposite. I'm open to suggestion on how this can be done.

**Comment by faceyspacey** [comment]
Hey @eXon . First off, congrats on that rediculously awesome webpack package. As for for your **BlazeToReact** package, it's basically doing a small piece of what **Blaze React** (this package) is attempting to do:

- **BlazeToReact** only handles conversion of the html code for one template at a time: https://github.com/thereactivestack/meteor-blazetoreact/blob/master/lib/BlazeToReact-client.jsx 

- **Blaze React** is a complete transpiler for blaze code beyond just html interpolation, including: events, helpers, lifecycle handlers (`onCreated` -> `componentWillMount` etc), dynamic inclusion of child templates/components, passing events down so they work on child templates, context binding throughout of all this, etc--all with the goal that your Blaze code will work as is by simply changing out the underlying renderer. 

What you end up with is React components. So it's just like using React. We happen to use the meteor plugin hook like u do, and they only support one per extension at time (i.e. `.js`) --so all i was saying was that ur plugin code and our plugin code will have to coexist since webpack and react are commonly used together. Maybe you thought we were just handling the .html extension, but because of parsing events, helpers, and callbacks, we gotta parse the .js extension as well. 

Maybe Meteor will support multiple compilers able to compile the same language in succession. Not sure how the order would be specified--guess by the order the package was added. Not sure what side-effects could be introduced. I get the feeling it won't come any time soon. So we're not gonna wait on MDG. 

... That all said, your `BlazeToReact()` function may be more powerful than it looks at first glance. You do basically lose any performance enhancements React provides since it's basically rendering all of blaze in React. I'm assuming if you render a template, its child templates will render with the correct context, and even all the helpers working reactively. Hell, the events will even work. But it's basically like a Blaze *iframe* within a React site. The real question is would that work within React Native. And that's where it would no longer hold-up. The reason being that Blaze style reactivity doesn't support iOS or Android. The views wouldn't update. Even if they werent reactive, and only needed to be rendered once from the top down, i'm not even sure the blaze HTML would holdup as native views. In addition, React Native is just one tool from the React ecosystem we couldn't take advantage of. The whole purpose of *Blaze React* is to gain access to the React ecosystem--you don't get that by rendering Blaze in the root component. And rendering bits and pieces of components like *loginButtons* isn't what we are going for. That's a stop gap for people that wanna use Blaze packages right now without this transpiler. When this transpiler is rock solid, all those blaze packages will naturally work in React. We have a few major challenges, and we need as much help as we can get, but things are looking like it's all gonna be possible based on my research.
